pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+}
+
+func NewAppManager(amProtocol interfaces.ApplicationManagementProtocol) *AppManager {
+	return &AppManager{
+		amProtocol: amProtocol,
+	}
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Name() string {
+	return constants.AppManagerHandlerName
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) ServiceInit() error {
+	return nil
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Start() error {
+	appMgr.controller = NewController(appMgr)
+	appMgr.controller.Run()
+	return nil
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Stop() {
+	// noop
+}
+
+// handle the updates from the scheduler and sync the status change
+func (appMgr *AppManager) HandleCallbackEvents() func(obj interface{}) {
+	return func(obj interface{}) {
+		if event, ok := obj.(events.ApplicationEvent); ok {
+			if events.AppStateChange == event.GetEvent() {
+				if shimEvent, ok := event.(shimcache.ApplicationStatusChangeEvent); ok {
+					appID := event.GetApplicationID()
+					log.Logger.Info("Status Change callback received",
+						zap.String("app id", appID),
+						zap.String("new status", shimEvent.State))
+					var app = appMgr.amProtocol.GetApplication(appID).(*shimcache.Application)
+					appCRD, err := appMgr.controller.lister.Applications(app.GetTags()[constants.AppTagNamespace]).Get(appID)
+					if err != nil {
+						log.Logger.Warn("Failed to query app CRD for status update",
+							zap.String("Application ID", appID),
+							zap.Error(err))
+						return
+					}
+					crdState := convertShimAppStateToAppCRDState(shimEvent.State)
+					if crdState != "Undefined" {
+						appMgr.updateAppCRDStatus(appCRD, crdState)
+					} else {
+						log.Logger.Error("Invalid status, skip saving it",
+							zap.String("App id", appID))
+					}
+				}
+			}
+		}
+	}
+}
+
+type Controller struct {
+	kubeClient       kubernetes.Interface
+	extensionsClient apiextensionsclient.Interface
+	appClient        applicationclient.Interface
+	informer         cache.SharedIndexInformer
+	lister           applister.ApplicationLister
+}
+
+func NewController(appMgr *AppManager) *Controller {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		log.Logger.Fatal(err.Error())
+	}
+
+	kubeClient := kubernetes.NewForConfigOrDie(config)
+	apiextensionsClient := apiextensionsclient.NewForConfigOrDie(config)
+	appClient := applicationclient.NewForConfigOrDie(config)
+
+	informerFactory := appinformers.NewSharedInformerFactory(appClient, time.Minute*1)
+	informer := informerFactory.Apache().V1alpha1().Applications()
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    appMgr.addApp,
+		UpdateFunc: appMgr.updateApp,
+		DeleteFunc: appMgr.deleteApp,
+	})
+	informerFactory.Start(wait.NeverStop)
+
+	utilruntime.Must(appv1.AddToScheme(appscheme.Scheme))
```

Comment:
what is this for?

Reply:
this utilruntime.Must call is handling the errors. Actually it will panic if the scheme registration is failing. I've seen this pattern used in other places, when starting the controller, so I followed this one as well.

Reply:
Got it, thx